### PR TITLE
Detection of curled edges to enhance slowdown for overhangs algorithm

### PR DESCRIPTION
**Pull request by igiannakas**
**"Prusa slicer slows down to the lowest speed (20mm/sec) on the tip of the bow even though the overlap is 75% or their abouts"**. 
This is due to the detection of a potential edge that may curl up due to the angle of the extrusion.

This PR aims to replicate this functionality in Orca slicer. 

It has been implemented as a parameter and is by default switched off as
1. For continuation of user experience from the current version
2. It is no longer necessary to be as aggressive with the slow down on overlapping perimeters as the algorithm takes into account where curls happen. It therefore may benefit that the user increases the minimum speed for the 25-50 and 50-75 overlap speeds
3. It needs the overhang perimeters to be printed using the overhang speed instead of the bridge speed

**Current Orca:**
Screenshot 2023-09-11 at 16 28 30

Can clearly see the deformation in the hull because the slowdown was not enough, even though the minimum speed was set to 20mm/sec
IMG_2392

**This PR:**
Screenshot 2023-09-11 at 21 18 15
Screenshot 2023-09-11 at 21 19 20

**Prusa Slicer:**
Screenshot 2023-09-11 at 16 32 09

### CODE REVIEW

File: src/libslic3r/GCode.cpp [hunk 1]
```diff
@@ -4545,8 +4545,13 @@ std::string GCode::_extrude(const ExtrusionPath &path, std::string description,
              (m_config.get_abs_value("overhang_4_4_speed") < 0.5) ?
                  FloatOrPercent{100, true} :
                  FloatOrPercent{m_config.get_abs_value("overhang_4_4_speed") * 100 / out_wall_ref_speed, true},
-             FloatOrPercent{m_config.get_abs_value("bridge_speed") * 100 / out_wall_ref_speed, true},
-             FloatOrPercent{m_config.get_abs_value("bridge_speed") * 100 / out_wall_ref_speed, true}});
+             (m_config.get_abs_value("overhang_4_4_speed") < 0.5) ?
```
Comment (SoftFever) [inline src/libslic3r/GCode.cpp:4548]: Hi @igiannakas 
This is behavor is intentional.
If a perimeter is not more than 90 percents supported, it is considered as bridge.
If you look at the overhang speed setting UI, it mentions this.

Comment (SoftFever) [reply src/libslic3r/GCode.cpp:4548]: As of the 13 percents here in the code, this is the magic number I come up with afte comparing/tuning many many models.

Comment (igiannakas) [reply src/libslic3r/GCode.cpp:4548]: Understood. Should we explore the opportunity to scan for curled edges and use that as a mechanism to further refine the slow down for overhangs algorithm? What do you think?

**Comment by igiannakas** [comment]
OK so I've reverted the handling of overhang perimeters. Started code prep to add identification of curled edges to experiment. PR is not ready to merge as it doesnt do anything yet, but I'll keep synching to the branch while experimenting with it.

Hope that's ok?

File: src/libslic3r/GCode.cpp (hunk 1 above)
Comment (SoftFever) [reply src/libslic3r/GCode.cpp]: Sure! 😀

**Comment by igiannakas** [comment]
@SoftFever  submitted first prototype of the code. It compiles, functions and indeed does slow down the perimeters where curled edges are possible.

However, I will need some assistance if possible in debugging it. It doesnt appear to have as much of an effect in the slow down as I had hoped for - at least not close to the massive slow down that I see in Prusa slicer. 

Maybe I'm doing something wrong here; it was quite a challenge to merge this in from the latest Prusa Slicer as the BBL code base was quite old.

**Comment by igiannakas** [comment]
**Progress update:**
Code is now finally working - external perimeters are slowed down as needed when there is likelyhood of curling up. Please compile locally @SoftFever and take a look :) 

ToDo:
1. Reslicing the model causes the algorithm not to run - need to find where this fails to trigger the call for this.
2. Slowdown of internal perimeters not working yet.
3. Had to remove the option to use bridge speed for overhang slow down. That is because it would actually speed up in areas where curling was happening because it was using the parameter for bridging.
4. Ideally we could set this up as an option. Turn it off and the default behaviour is present. Turn it on and the overhang speed is used for overhang perimeters and additional slow down is introduced at curled edges.

**Comment by igiannakas** [comment]
Hi @SoftFever 

I've matured this to the most my skills allow - I would appreciate your help if you can please with the below:

The algorithm runs correctly the first time a model is loaded and sliced but then if any changes to the model are made (i.e. layer height), it doesnt run.

It appears that the `this->set_started(posEstimateCurledExtrusions)` function call on estimate_curled_extrusions function returns false. 

I cannot find for the life of me where these are set to "true" in the code. I presume it is to do with threading and some orchestrator somewhere is kicking these off, but cannot find where this is.

Would love your help if you can?

All the best

**Comment by igiannakas** [comment]
@SoftFever this is now code complete and ready for review. 

Keen to hear your feedback and potentially any builds/additions/extensions.

All the best!

**Comment by igiannakas** [comment]
> @igiannakas thank you so much! I will only have time to do proper code review and testing after v1.7.0 release. Meanwhile, I will enable the CICD for this PR so that others can help testing and validating the result. Again, really appreciate your contribution!

Your welcome! It was fun working on that - taught me a hell of a lot on the code structure off orca and prusa slicer. Looking forward to the next!

**Comment by HahBikes** [comment]
Hey @igiannakas, Great work on this! I was just wondering if it was also possible to use the **"fan speed for overhangs"** in the slowed-down sections even for the inner walls? That way along with the slowdown and force cooling, it will really help the curled corners settle down.

**Comment by YanceyA** [comment]
> Hey @igiannakas, Great work on this! I was just wondering if it was also possible to use the **"fan speed for overhangs"** in the slowed-down sections even for the inner walls? That way along with the slowdown and force cooling, it will really help the curled corners settle down.

This would need to be an option as some cases you don't want to slow down and use a (larger) fan speed on overhangs. For ABS this could actually promote edge and corner cooling from the faster cooling rates.

**Review by SoftFever — APPROVED** [review]
It looks all good to me.
Great work!

**Comment by SoftFever** [comment]
Merged.
Cheers 👍

**Comment by igiannakas** [comment]
Thank you 🙏🏻

**Comment by sleewok** [comment]
Is this supposed to slow down the inner perimeters as well? I am not seeing that. See in my image that only the outer perimeter is slowing down?

image

**Comment by igiannakas** [comment]
> Is this supposed to slow down the inner perimeters as well? I am not seeing that. See in my image that only the outer perimeter is slowing down?
> 
> image

Can you post your 3mf project file please?

**Comment by sleewok** [comment]
> > Is this supposed to slow down the inner perimeters as well? I am not seeing that. See in my image that only the outer perimeter is slowing down?
> > image
> 
> Can you post your 3mf project file please?

Here is a 3mf that exhibits the same issue:

[bltouchw-ASA2.zip](https://github.com/SoftFever/OrcaSlicer/files/13794249/bltouchw-ASA2.zip)

**Comment by igiannakas** [comment]
Not a bug - setting works as expected:

image


See the benchy next to it - the hull is slowing down even in the internal perimeters.

If I were you I'd use these settings to print it. You're going WAY too fast for such a small part so there is simply not enough time to cool it down, especially as there is no part cooling fan running as its ASA

image

**Comment by sleewok** [comment]
Okay, I see what you are saying. I am printing with a high speed ASA filament so it is a bit different from traditional.

This brings to mind the possibility of adjusting the threshold for angles that will slow down for inner perimeters. I could see this as being very beneficial for smaller parts. Just as we can set the overhang speed at varying percentages, it would be great to be able to toggle/force "slow down for curled perimeters" for each as well. Could be a boolean that is checked before running the analysis to see if it should be applied that would force it to be applied? I haven't looked at the code to see how difficult it would be. 

Thoughts?

**Pull request merged**